src/boot/java/io/github/brassmc/brassloader/boot/discovery/ModDiscovery.java

```diff
@@ -0,0 +1,260 @@
+        JsonValue itemsJson = object.get(name);
+        String[] items;
+        if(itemsJson != null && itemsJson.isArray()) {
+            String[] arr = itemsJson.asArray().values().stream().map(JsonValue::asString).toArray(String[]::new);
+            for (int i = 0; i < arr.length; i++) {
+                String author = arr[i];
+                lengthCheck(name + " " + (i + 1), author, modLoc, 4, 32);
+            }
+
+            items = arr;
+        } else if(itemsJson != null && itemsJson.isString()) {
+            items = new String[] { itemsJson.asString() };
+        } else
+            items = new String[0];
+
+        if(shouldThrowIfMissing)
+            if(items.length == 0)
+                throw new MetadataParseException("At least 1 " + name + " must be provided for mod: " + modLoc);
+
+        return items;
+    }
+
+    private static String[] getArrayOrString(JsonObject object, String name, Path modLoc) {
+        return getArrayOrString(object, name, modLoc, false);
+    }
+
+    private static void lengthCheck(String name, String value, Path modLoc, int minLength, int maxLength) {
+        if(value == null || value.isBlank())
+            throw new InvalidModidException("Provided " + name + "(" + value + ") in mod(" + modLoc + ") must not be blank!");
+        if(value.length() < minLength)
+            throw new InvalidModidException("Provided " + name + "(" + value + ") in mod(" + modLoc + ") must be at least " + minLength + " characters!");
+        if(value.length() > maxLength)
+            throw new InvalidModidException("Provided " + name + "(" + value + ") in mod(" + modLoc + ") must be fewer than " + maxLength + " characters!");
+    }
+
+    private static String getString(JsonObject object, String name, String defaultValue, Path path) throws MetadataParseException {
+        String value = object.getString(name, defaultValue);
+        if(value != null && value.isBlank()) {
+            throw new MetadataParseException(name + " was not provided for mod: " + path);
+        }
+
+        return value;
+    }
+
+    private static String getString(JsonObject object, String name, Path path) throws MetadataParseException {
+        return getString(object, name, "", path);
+    }
+
+    private static String validateURL(UrlValidator validator, String fieldName, String url, Path path) {
+        if(!url.isBlank() && !validator.isValid(url))
+            throw new MetadataParseException(fieldName + " url is invalid in mod: " + path);
+
+        return url;
+    }
+
+    private static boolean isValidModid(String modid) {
+        for(int index = 0; index < modid.length(); ++index) {
+            if (!modidAllows(modid.charAt(index))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static boolean modidAllows(char c) {
+        return c == '_' || c == '-' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '.';
```

Comment:
Please remove dashes as valid mod ID characters. JMPS does not allow dashes in module names, and the module of the mod jars will have the mod ID as their name. 
```suggestion
        return c == '_' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '.';
```